src/utilities/path.ts

```diff
@@ -14,6 +15,25 @@ export type Path<
   params: Identity<ExtractParamsFromPathString<T, P>>,
 }
 
-export function path<T extends string, P extends PathParams<T>>(_path: T, _params: Identity<P>): Path<T, P> {
-  throw 'not implemented'
+function getParam<P extends Record<string, OptionalParam>>(params: P, param: string): Param {
+  return params[param] ?? String
+}
+
+export function path<T extends string, P extends PathParams<T>>(path: T, params: Identity<P>): Path<T, P> {
+  const paramPattern = /:\??([\w]+)(?=\W|$)/g
+  const matches = Array.from(path.matchAll(paramPattern))
+
+  const paramAssignments = matches.map(([match, paramName]) => {
+    const isOptional = match.startsWith(':?')
+    const param = getParam(params, paramName)
+
+    return {
+      [paramName]: [isOptional ? optional(param) : param],
+    }
+  })
+
+  return {
+    path,
+    params: mergeParams(...paramAssignments) as Identity<ExtractParamsFromPathString<T, P>>,
```

Comment:
Would this work as well? Just cause its shorter and will automatically update if we change the Path type. Don't feel strongly about this just had the thought when I saw it. 
```suggestion
    params: mergeParams(...paramAssignments) as Path<T, P>["params"],
```